### DISCUSSION TITLE
Fan Control and Reading Issue with VirtualSMC.kext and SMCSuperIO.kext

### DISCUSSION THREAD

**Discussion by shiecldk** [Qonfused/ASUS-ZenBook-Duo-14-UX481-Hackintosh#34] [General]
With the corrected Kernel (kexts loading) sequence in OpenCore, on UX582, the iStat Menus is able the read the sensor data from SMCSuperIO.kext. However, sometimes the Fan Speed reading is wrong (showing ~16700 rpm); normally, it should be ~1400 rpm. Also, there despite the control options available inside Fan Rules in iStat Menus, it seems to have no effects on tweaking the fan speed, which leads me to suspect that we should look into the DSDT/SSDT or check the SMCSuperIO.kext.

<img width="335" alt="Screen Shot 2023-01-26 at 11 33 08 PM" src="https://user-images.githubusercontent.com/42949962/214878005-0adece37-8f9a-48bb-ad59-f1913c0a8e42.png">
<img width="271" alt="Screen Shot 2023-01-26 at 11 33 56 PM" src="https://user-images.githubusercontent.com/42949962/214878209-e37969d6-de92-48f1-af71-7d8877c103b9.png">

When the reading of fan speed goes wrong:
<img width="335" alt="Screen Shot 2023-01-27 at 1 27 20 AM" src="https://user-images.githubusercontent.com/42949962/214906460-48bca043-f766-4340-815a-3ee80dbbef80.png">

**Comment by Qonfused** [comment]
Not entirely sure that these laptops have supported superio chips; the kext is more so intended for discovering superio chips for desktop platforms. There are some instructions on mapping the embedded controller under the VirtualSMC docs [here](https://github.com/acidanthera/VirtualSMC/blob/master/Docs/EmbeddedControllers.md) which may warrant some investigation.

I did do some digging into the UX481FL's DSDT in this issue in #37 and found some EC methods + WMI methods for controlling fan speed + thermal policy:

> --- https://github.com/Qonfused/ASUS-ZenBook-Duo-14-UX481-Hackintosh/issues/37#issue-1583232550 ---
> This is mainly relevant for the UX481FA/FL since the UX581/UX582 models have a media key + native ACPI implementation for toggling thermal policy. For all models, both (2) fans are controlled by the BIOS + embedded controller, though cooling policy is still configurable via ACPI methods as hinted by the system control driver used by MyASUS.
> 
> RW-Everything on Windows should show what EC registers are actually being used for fan control, however there should be more standard ACPI methods called by the system control driver (which you should always use rather than writing to EC registers directly).
> 
> Below are some observations that may clue into the native ACPI implementation of this:
> 
> * An ACPI method [`\_TZ_.RTMP`](https://github.com/Qonfused/ASUS-ZenBook-Duo-14-UX481-Hackintosh/blob/e3976205a69052a45baa46c0fb8c42c9890a4b5f/resources/firmware/UX481FL/DSDT.dsl#L58852-L58867) reads and stores the current CPU temperature.
>       
>   * The `ECPU` EC method reads the current CPU temperature.
>       * This method is used under the `_TMP` method of the [`THRM` thermalzone](https://github.com/Qonfused/ASUS-ZenBook-Duo-14-UX481-Hackintosh/blob/e3976205a69052a45baa46c0fb8c42c9890a4b5f/resources/firmware/UX481FL/DSDT.dsl#L58775-L58845), which should shed some light on how cooling policy is configured.
> 
>   * The [`\_TZ_.RFAN`](https://github.com/Qonfused/ASUS-ZenBook-Duo-14-UX481-Hackintosh/blob/e3976205a69052a45baa46c0fb8c42c9890a4b5f/resources/firmware/UX481FL/DSDT.dsl#L58899-L58927) method (in the same scope) is responsible for reading the fan speed.
>       
> * The `ECAV` EC method checks if the embedded controller is available/ready, otherwise the method aborts with value zero.
>   * The `ST83` EC method outputs byte values for a given fan idx (range appears to be between 0x00 and 0xFF)
>   * The `TACH` EC method outputs the RPM from a tachometer for a given fan idx (appears to max at 6000 RPM).
> 
> * Potentially relevant, under WMND (IIA0=0x00110019) there are several blocks checking a `FANF` variable.
>       
>   * This is likely referenced in the MyASUS driver for fan diagnostics; it potentially refers to fan PWM frequency.

> --- https://github.com/Qonfused/ASUS-ZenBook-Duo-14-UX481-Hackintosh/issues/37#issuecomment-1430561014 ---
> It seems that the `ST98` EC method controls the maximum allowed fan RPM; technically this is how fan speed is actually changed.

> --- https://github.com/Qonfused/ASUS-ZenBook-Duo-14-UX481-Hackintosh/issues/37#issuecomment-1434030682 ---
> [CPU fan control](https://github.com/torvalds/linux/blob/3ac88fa4605ec98e545fb3ad0154f575fda2de5f/include/linux/platform_data/x86/asus-wmi.h#L81) (WMND, IIA0=0x00110013):
> 
> ```asl
> 55455 | If ((IIA0 == 0x00110013))
> 55456 | {
> 55457 |     Local0 = ^^PCI0.LPCB.EC0.RRAM (0xCC, 0x30)
> 55458 |     If ((IIA1 == Zero))
> 55459 |     {
> 55460 |         Local1 = (Local0 & 0xFFFFFFFFFFFFFFBF)
> 55461 |     }
> 55462 |     ElseIf ((IIA1 == One))
> 55463 |     {
> 55464 |         Local1 = (Local0 | 0x40)
> 55465 |     }
> 55466 | 
> 55467 |     ^^PCI0.LPCB.EC0.WRAM (0xCD, 0x30, Local1)
> 55468 |     Return (One)
> 55469 | }
> ```
> 
> ^^ This appears to toggle max fan rpm.
> 
> (WMND, IIA0=0x00110019):
> 
> ```asl
> 55176 | If ((IIA0 == 0x00110019))
> 55177 | {
> 55178 |     If ((FANF == One))
> 55179 |     {
> 55180 |         Return (0x00010001)
> 55181 |     }
> 55182 |     Else
> 55183 |     {
> 55184 |         Return (0x00010000)
> 55185 |     }
> 55186 | }
> ...
> 55428 | If ((IIA0 == 0x00110019))
> 55429 | {
> 55430 |     FANL (IIA1)
> 55431 |     Return (One)
> 55432 | }
> ```
